### ISSUE TITLE
About acc performance on CIFAR-10 and training time.

### ISSUE THREAD

**Issue by zlannnn** [EmilienDupont/augmented-neural-odes#2]
Hi guys, your works and codes are great.
However, I am a little confused about why there were no accuracy comparsion between NODE and your ANODE on CIFAR-10 since you did tests with both methods.
I am trying to do the test on CIFAR-10, and I noticed that training time per epoch of ANODE increasing very fast rather than NODE, is that how things should going on with ANODE?

Regards

**Comment by EmilienDupont** [comment]
Hi, thanks for your comments!
Since we were mostly interested in showing to which extent NODEs and ANODEs could fit various datasets, we only included plots of the losses since this is what we were optimizing for. However, several people have asked for test accuracy measurements too, so we will include those in an updated version of the paper.
And the behaviour you see when training on CIFAR-10 is expected. Indeed, since ANODEs are easier to optimize, they learn a better approximation of the mapping from images to features in fewer iterations than NODEs. Since a better approximation of this function typically implies that the resulting ODE becomes more difficult to solve, the NFEs will increase faster for the ANODE (this is also what you can see in Figure 10 in the paper, the NFEs increase at roughly the same rate). However, the ANODE should also have a much lower loss than the NODE. If you wait until the NODE achieves the same loss as the ANODE, the NODE should have much higher NFEs. This is one of the main observations of the paper: to learn an equally complex function (i.e. to achieve an equal loss on a dataset), NODEs require a more complicated flow (i.e. they require more time to solve the ODE).
Hope that makes sense and let me know if you have any more questions!

**Comment by zlannnn** [comment]
Hi. Thanks for your reply, that explained a lot.
The only major issue confused me is that the training time per epoch of your ANODE increased much faster than NODE while I trying to test acc on CIFAR-10.
I knew training time vs epoch number in NODE have a linear relationship and your work achieved better loss. 
I was trying to compare NODE and your ANODE under same condition, all start with around 100s at first epoch, training time per epoch of NODE was increasing slowly while ANODE went to 200s in just 15-20 epochs.
I am not sure the issue above is due to my code or your model, that is the reason why I opened this issue.
Looking for your reply and your new version paper.

**Comment by EmilienDupont** [comment]
Hi, thanks for your reply! I have a couple of questions which can hopefully help us figure out what's happening:

1. Are you running the code in this repo or your own version of the code? If you are not running the code from this repo, could you try running it and see if you still observe the same difference?
2. Are you running on GPU or CPU?
3. Do you have some data/plots for the NFEs and number of seconds per epoch when you train both with NODEs and ANODEs?

If the NFEs are roughly the same for NODEs and ANODEs but you still observe a difference in wall clock time, then it could be that ANODEs run slightly slower because they can have more parameters than NODEs (depending on how much you augment). However, in our experiments this difference was very small, so I'm not sure if this is the issue.

**Comment by zlannnn** [comment]
Hi!
1. I was calling ANODE by " anode = ConvODENet(device,img_size=(3,32,32),num_filters=args.filters,output_dim=10,time_dependent=True,non_linearity='relu',augment_dim=args.augments,tol=args.tol,adjoint=True)"
2. Where device is GPU. I used filters = 64, augments =10 as you recommended in the paper for cifar-10 and tol =0.1 for faster training.
3.  I just changed optimizer from adam to SGD and most the issue is gone. The training time increasing much slower which is understandable.
However, the loss suddenly increase and the program stucked at epoch 19 while lr = 0.001. If the lr is larger, the situation would happen eailer.
![BOOB](https://user-images.githubusercontent.com/15451867/61851530-366de080-aef2-11e9-9b95-6b7d7a35ebb1.png)
And I also a little confused that why the test acc(red one in figure) is always larger than training acc.
![test](https://user-images.githubusercontent.com/15451867/61852088-8ef1ad80-aef3-11e9-9814-7eddf11bfa8c.png)

Although the situation is much better now, there are still seems something wrong.
Comparing ANODE and ODE before it stucked(18 epoch, **ANODE blue and ODE red** in following figures.)
For acc.
![acccc](https://user-images.githubusercontent.com/15451867/61853529-142a9180-aef7-11e9-8be7-490ec6650a4e.png)
For NFEs forward
![NFESf](https://user-images.githubusercontent.com/15451867/61853741-974be780-aef7-11e9-88cc-737fc31a233e.png)
For NFEs backward
![NFEsB](https://user-images.githubusercontent.com/15451867/61853791-b34f8900-aef7-11e9-8718-4cf513e315bc.png)
For training loss
![train loss](https://user-images.githubusercontent.com/15451867/61853867-d9752900-aef7-11e9-8a47-d2f4fe1b81c3.png)
For test loss
![testloss](https://user-images.githubusercontent.com/15451867/61853936-f7428e00-aef7-11e9-94fd-f4659fad5323.png)
training time per epoch for ANODE is roughly from 132s to 155s.
for ODE is roughly from 53s to 54s.
All 64 fillters,SGD, lr 0.001, same bn bs all same setting. The only major difference is that Tianqi in ODE used a residual block in MNIST example so I also used it.
If change residual block.(Add no residual block ODE in figure as Yellow)
For training loss forward and back
![lossplus](https://user-images.githubusercontent.com/15451867/61856799-d1b88300-aefd-11e9-98a1-76cb836f922d.png)
![backplus](https://user-images.githubusercontent.com/15451867/61856909-0fb5a700-aefe-11e9-8fe1-70bebe452ac7.png)
And acc.
![accplus](https://user-images.githubusercontent.com/15451867/61857040-55726f80-aefe-11e9-8dbc-c20f66d25e9d.png)
And NFE.
![NFE](https://user-images.githubusercontent.com/15451867/61857082-658a4f00-aefe-11e9-9c39-e80d5af7d892.png)
A little worse but did not explain the issue.

I think your paper is the best after ODE. There must be something wrong in code but I really can not figure it out. The code works fine on ODE.
Here are the log files and the train.py I used.
[log&train.py.tar.gz](https://github.com/EmilienDupont/augmented-neural-odes/files/3430231/log.train.py.tar.gz)
I also have some tiny results related to ODE. 
Would you like give the draft paper a quick review next month?  Really appreciate for your attention :)

**Comment by EmilienDupont** [comment]
Hi again! Thanks for the detailed reply :)

I've also run experiments on my side and using the default parameters with ANODE I get to about 98% test accuracy in 3-4 epochs on MNIST, whereas it takes roughly 20 epochs for NODEs to get to 96% accuracy.

ANODE (train in red, test in blue)
<img width="378" alt="Screen Shot 2019-07-29 at 2 15 12 PM" src="https://user-images.githubusercontent.com/12076034/62083260-65929380-b20b-11e9-895e-a53833ff9e10.png">

NODE (train in red, test in blue)
<img width="375" alt="Screen Shot 2019-07-29 at 2 15 17 PM" src="https://user-images.githubusercontent.com/12076034/62083271-6aefde00-b20b-11e9-9822-0d0b987d6cd6.png">

The reason I think you are not getting good results is because you are setting tol=0.1. Since pixel values are in [0, 1], this implies that the ODE solver is okay with an error of 10% on average for each pixel value. This is a very big error, e.g. this is a CelebA image with and without noise on that scale:

<img width="150" alt="celeba_example" src="https://user-images.githubusercontent.com/12076034/62083870-c66e9b80-b20c-11e9-87ce-07c312f8ba72.png">
<img width="150" alt="celeba_example_noisy" src="https://user-images.githubusercontent.com/12076034/62083889-ce2e4000-b20c-11e9-935f-b225c5a5ff13.png">

Can you try setting tol=1e-3 (or smaller) and see if you get better results? Thanks!

**Comment by zlannnn** [comment]
Hi! Thanks for your attention
I tried lower tol and switched cifar to minist, results are getting better but the problem still remaining.
(I used augment 3 and 5. 0.01 lr) 

![train loss](https://user-images.githubusercontent.com/15451867/62364210-e6c78e80-b55b-11e9-927c-abbab1077c61.png)
![train acc](https://user-images.githubusercontent.com/15451867/62364218-eaf3ac00-b55b-11e9-9f2b-08cdea7c391c.png)
![test acc](https://user-images.githubusercontent.com/15451867/62364221-ecbd6f80-b55b-11e9-92c0-972f25daad9e.png)

Wondering what augment number did you use for the above results?
For ODE that you used for compare, did that model include two resblocks as Tianqi's minist example?
I used ( 'adams': VariableCoefficientAdamsBashforth) solver, I guess you may used Dopri5Solver, is that true?
Sorry for so many questions, I really want verify something on your augmented ODE cause it's the best updated ODE now.

**Comment by EmilienDupont** [comment]
Hi! For the above results I used aug 5. I've included the full config file I used below. I also used the `dopri5` solver for both NODE and ANODE experiments. In terms of learning rate I used `0.001` which is lower than the `0.01` you used, although I don't think this should make a big difference (it looks like you also decrease the learning rate later in training from your training loss plot).

Could you try running your experiments using the same params described in the json below? You should get the same results as me in that case.

From looking at your loss and accuracy plots, I'm surprised by how slowly the loss decreases/the accuracy increases. Usually, you should see very good performance after 3-4 epochs already (at least with ANODEs). In your train loss plots, it seems like this takes about 20 epochs.

Another weird thing is that it seems like NODEs get to about 98% test accuracy in 1 epoch. This seems surprising given how long it takes for the training loss to decrease.

I also did not add two resnet blocks as in the original NODE paper. If we include two resnet blocks, it becomes harder to compare the performance of NODEs and ANODEs because most of the modeling may be done in the resnet blocks. Further, if you use resnet blocks you also lose most of the advantage of ODE models (invertibility, querying the hidden state at any timestep, reduced number of paramters, easy jacobian computation in NFs etc). Could you try running your experiments without the resnet blocks and see if you can reproduce the results I have? If you can reproduce them, you should try to add the resnet blocks back. I suspect the difference in performance between NODEs and ANODEs would be reduced in this case since the resnet blocks could break the homeomorphism constraints which NODEs have.

```
{
 "id": "mnist_acc", 
 "num_reps": 2, 
 "dataset": "mnist", 
 "model_configs": [{ 
         "type": "anode", 
         "num_filters": 64, 
         "augment_dim": 5, 
         "time_dependent": true,
         "lr": 0.001, 
         "non_linearity": "relu", 
         "weight_decay": 0.0, 
         "validation": true
         }, 
         {
         "type": "odenet", 
         "num_filters": 92, 
         "time_dependent": true, 
         "lr": 0.001, 
         "non_linearity": "relu", 
         "weight_decay": 0.0, "validation": true
 }], 
 "training_config": {"batch_size": 256, "record_freq": 10, "print_freq": 10, "epochs": 20}
}
```

**Comment by zlannnn** [comment]
Thanks for your reply.
I will try that！

**Comment by zlannnn** [comment]
Hi!
Actually NODE with resblocks get to about 98% test accuracy in 1 epoch because high lr.
Without resblocks, ANODEs are much better.
BTW is it possible for ANODEs training under a higer lr like 0.1 in your experiments?
Wondering that cause it always cause problems here.

**Comment by EmilienDupont** [comment]
Hi, glad to see you got the same results as us when removing the resnet blocks! Including resnet blocks typically means that most of the modeling ends up being done by the resnet blocks and removing the NODE layer in those models barely decreases performance.

I haven't experimented with using higher learning rates, so I'm not sure how well it would work. We didn't really play around with this too much, but it might be possible to get faster convergence with a higher learning rate or a better learning rate schedule.

**Comment by zlannnn** [comment]
On mnist dataset it's true that removing NODE layer barely decreases performance. For dataset like cifar 10, there are about 3% difference w/wo NODE layer according my test.

I also think that should get faster convergence but loss exploded before model trained while lr = 0.1. Not sure it's my issue or not. I checked code few times but have no clue with it.

**Comment by EmilienDupont** [comment]
Nice! I only did the resnet block experiments on MNIST, interesting to see the difference is larger for CIFAR-10.

I haven't had a chance to try training with a learning rate of 0.1 yet. However, it does seem to be a very high learning rate, so I'm not too surprised the loss might explode. I don't think there is an issue with your code, I think the model just needs to use a lower learning rate.

**Comment by zlannnn** [comment]
It's quite interesting cause ODE related nets sometimes require different FLOPs via different lr (not that much difference, but standard nets do not).

BTW, does your Conv2dTime count as one extra dimension?  It does take time into consideration as third, but NODEs did not get rid of its weakness through time-dependent Conv via your demo.

**Comment by EmilienDupont** [comment]
No time does not count as an extra dimension. Indeed, adding time just means the input to the ODE function has one extra dimension, it doesn't mean that the hidden state (or the initial condition of the ODE) has an extra dimension. Indeed, if `h(t)` is in `R^d` and it is evolved according to `dh(t)/dt = f(h(t), t)`, then adding time just means that `f(h(t), t)` maps from `R^(d+1) -> R^d`. But the hidden state `h(t)` (and input/output) of the ODE are still in `R^d`, so this doesn't actually remove the weaknesses discussed in the paper.

**Comment by zlannnn** [comment]
Thanks a lot!
I tried Chen's ODE-Net like structure on your 2d example.  NODEs with no extra dimension but Res structure could sometime achieve input space figures like this if time also a part of the input.
![stable](https://user-images.githubusercontent.com/15451867/66497385-5e371280-eaf7-11e9-92e4-16463c79b158.png)
It not as good as ANODEs does, but quite close.
The inner shape seems to have edges because of discrete points  ODE-Net combined with.

**Comment by EmilienDupont** [comment]
That's really cool! Thanks for sharing.
This behaviour is indeed expected since the ResNet blocks can break the homeomorphism constraint of Neural ODE layers and so can separate the regions. Indeed, you get similar results to the one you have when just using a ResNet (with no ODE layers).
